internal/backend/sftp/sftp.go

```diff
@@ -258,23 +258,40 @@ func (r *SFTP) Save(ctx context.Context, h restic.Handle, rd restic.RewindReader
 	}
 
 	filename := r.Filename(h)
+	dirname := r.Dirname(h)
 
-	// create new file
-	f, err := r.c.OpenFile(filename, os.O_CREATE|os.O_EXCL|os.O_WRONLY)
+	// Ensure that the containing directory exists.
+	if err := r.c.MkdirAll(dirname); err != nil {
+		debug.Log("error creating dir %v: %v", r.Dirname(h), err)
+		return errors.Wrap(err, "Mkdir")
+	}
 
-	if r.IsNotExist(err) {
-		// error is caused by a missing directory, try to create it
-		mkdirErr := r.c.MkdirAll(r.Dirname(h))
-		if mkdirErr != nil {
-			debug.Log("error creating dir %v: %v", r.Dirname(h), mkdirErr)
-		} else {
-			// try again
-			f, err = r.c.OpenFile(filename, os.O_CREATE|os.O_EXCL|os.O_WRONLY)
-		}
+	// Try and see if there is space left for this file.
+	// We do this because the OpenSSH sftp-server only returns
+	// a generic "failure" code when the filesystem is full,
+	// so detecting after the fact is impossible.
+	if err := r.checkSpace(dirname, rd.Length()); err != nil {
```

Comment:
This has a similar performance problem as `MkdirAll`. `checkSpace` is expected to nearly always return that there is still enough free space. So why not try creating the file first and if that fails, just check whether we ran out of space?

Reply:
Ok, did that instead.